### PR TITLE
[AWS] [S3] add generators to spec

### PR DESCRIPTION
**Pull request by nagypeterjob**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

**What is the current behavior? (You can also link to an open issue here)**

**What is the new behavior (if this is a feature change)?**

**Additional info:**

* [x] cp
* [x] ls
* [x] mb
* [x] mv
* [x] presign
* [x] rb
* [x] rm
* [x] sync
* [x] website

### CODE REVIEW

File: dev/aws/s3.ts [hunk 1]
```diff
@@ -1,3 +1,434 @@
...
+  "us-east-1",
+  "us-east-2",
+  "us-west-1",
+  "us-west-2",
+];
+
+const _prefix_s3 = "s3://";
+const _prefix_file = "";
+const _prefix_fileb = "fileb://";
+
```
Comment (QuiiBz) [inline dev/aws/s3.ts:29]: I would prefer camel case for variable names.

File: dev/aws/s3.ts [hunk 2]
```diff
@@ -1,3 +1,434 @@
...
+      var whatHasUserTyped = tokens[tokens.length - 1];
+
+      // Do not show file suggestions when s3:// typed
+      if (whatHasUserTyped.startsWith(_prefix_s3)) {
+        return "";
+      }
+
+      var folderPath = "";
+
+      var lastSlashIndex = whatHasUserTyped.lastIndexOf("/");
```
Comment (QuiiBz) [inline dev/aws/s3.ts:43]: Please use `const` or `let` (when you need to re-assign it later) for variables.

File: dev/aws/s3.ts [hunk 3]
```diff
@@ -1,3 +1,434 @@
...
+
+      return final_array;
+    },
+
+    trigger: (newToken, oldToken) => {
+      if (!newToken.startsWith(_prefix_file)) {
+        if (!oldToken) return false;
+
+        if (oldToken.startsWith(_prefix_file)) return true;
+        return false;
```
Comment (QuiiBz) [inline dev/aws/s3.ts:110]: Can be simplified:
```suggestion
        return oldToken.startsWith(_prefix_file));
```

File: dev/aws/s3.ts [hunk 4]
```diff
@@ -1,3 +1,434 @@
...
+      }
+
+      if (newToken.lastIndexOf("/") !== oldToken.lastIndexOf("/")) {
+        return true;
+      } else return false;
```
Comment (QuiiBz) [inline dev/aws/s3.ts:115]: Same:
```suggestion
      return newToken.lastIndexOf("/") !== oldToken.lastIndexOf("/"));
```

File: dev/aws/s3.ts [hunk 5]
```diff
@@ -1,3 +1,434 @@
...
+        if (whatHasUserTyped.startsWith("~/"))
+          folderPath = whatHasUserTyped.slice(0, lastSlashIndex + 1);
+        else if (whatHasUserTyped.startsWith("/")) {
+          if (lastSlashIndex === 0) folderPath = "/";
+          else folderPath = whatHasUserTyped.slice(0, lastSlashIndex + 1);
+        } else folderPath = whatHasUserTyped.slice(0, lastSlashIndex + 1);
+      }
+
+      return baseLSCommand + folderPath;
+    },
```
Comment (QuiiBz) [inline dev/aws/s3.ts:156]: This script is repeated 3 times and is almost identical. I believe we can simplify and make it more generic.

File: dev/aws/s3.ts [hunk 6]
```diff
@@ -1,3 +1,434 @@
...
+
+      return final_array;
+    },
+
+    trigger: (newToken, oldToken) => {
+      if (!newToken.startsWith(_prefix_fileb)) {
+        if (!oldToken) return false;
+
+        if (oldToken.startsWith(_prefix_fileb)) return true;
+        return false;
```
Comment (QuiiBz) [inline dev/aws/s3.ts:214]: Same:
```suggestion
        return oldToken.startsWith(_prefix_fileb));
```

File: dev/aws/s3.ts [hunk 7]
```diff
@@ -1,3 +1,434 @@
...
+      }
+
+      if (newToken.lastIndexOf("/") !== oldToken.lastIndexOf("/")) {
+        return true;
+      } else return false;
```
Comment (QuiiBz) [inline dev/aws/s3.ts:219]: Same:
```suggestion
      return newToken.lastIndexOf("/") !== oldToken.lastIndexOf("/"));
```

File: dev/aws/s3.ts [hunk 8]
```diff
@@ -1,3 +1,434 @@
...
+        if (!oldToken) return false;
+
+        if (oldToken.startsWith(_prefix_s3)) return true;
+        return false;
+      }
+
+      if (newToken.lastIndexOf("/") !== oldToken.lastIndexOf("/")) {
+        return true;
+      } else return false;
+    },
```
Comment (QuiiBz) [inline dev/aws/s3.ts:375]: Looks like this function is also repeated few times, we should extract it and use it everywhere instead of copy/pasting.

**Review by QuiiBz — CHANGES_REQUESTED** [review]

**Comment by nagypeterjob** [comment]
Hey @QuiiBz 

I do agree with most part of your review & I've addressed them where it made sense. Hope you will like it!

File: dev/aws/s3.ts [hunk 9]
```diff
@@ -1,3 +1,338 @@
...
+  let folderPath = "";
+  const lastSlashIndex = whatHasUserTyped.lastIndexOf("/");
+
+  if (lastSlashIndex > -1) {
+    if (whatHasUserTyped.startsWith("~/"))
+      folderPath = whatHasUserTyped.slice(0, lastSlashIndex + 1);
+    else if (whatHasUserTyped.startsWith("/")) {
+      if (lastSlashIndex === 0) folderPath = "/";
+      else folderPath = whatHasUserTyped.slice(0, lastSlashIndex + 1);
+    } else folderPath = whatHasUserTyped.slice(0, lastSlashIndex + 1);
```
Comment (cstrnt) [inline dev/aws/s3.ts:39]: I would love to have some curly brackets here to create blocks and make the code more readable

File: dev/aws/s3.ts [hunk 10]
```diff
@@ -1,3 +1,338 @@
...
+  };
+
+  const alphabeticalSortFilesAndFolders = (arr) => {
+    const dots_arr = [];
+    const other_arr = [];
+
+    arr.map((elm) => {
+      if (elm.toLowerCase() == ".ds_store") return;
+      if (elm.slice(0, 1) === ".") dots_arr.push(elm);
+      else other_arr.push(elm);
```
Comment (cstrnt) [inline dev/aws/s3.ts:69]: same here! I'm really big on blocks 😆

File: dev/aws/s3.ts [hunk 11]
```diff
@@ -1,3 +1,338 @@
...
+      ...dots_arr.sort(sortFnStrings),
+    ];
+  };
+
+  const temp_array = alphabeticalSortFilesAndFolders(arr);
+
+  const final_array = [];
+
+  temp_array.forEach((item) => {
+    if (!(item === "" || item === null || item === undefined)) {
```
Comment (cstrnt) [inline dev/aws/s3.ts:91]: You can actually do `item != null` which checks for null **and** undefined at the same time ;)

File: dev/aws/s3.ts [hunk 12]
```diff
@@ -1,3 +1,338 @@
...
+
+    return oldToken.startsWith(prefix);
+  }
+
+  return newToken.lastIndexOf("/") !== oldToken.lastIndexOf("/");
+};
+
+const _prefixS3 = "s3://";
+const _prefixFile = "";
+const _prefixFileb = "fileb://";
```
Comment (cstrnt) [inline dev/aws/s3.ts:121]: This is just some personal preference but I like making actual `consts` in Uppercase like `PREFIX_S3` do mark them as global, unchangeable consts

File: dev/aws/s3.ts [hunk 13]
```diff
@@ -1,3 +1,338 @@
...
+        // Parsing S3 CLI Output here
+        // Do this in a try block beacuse of the indexing magic
+        try {
+          //Example Output:
+          //                             PRE charts/
+          //  2021-05-08 10:15:53      81765 img.jpg
+          //
+          // After we have found at least 1 PRE keyword
+          // we can assume that all lines without PRE are files
+          if (parts[1] == "PRE") {
```
Comment (cstrnt) [inline dev/aws/s3.ts:244]:
```suggestion
          if (parts[1] === "PRE") {
```

File: dev/aws/s3.ts [hunk 14]
```diff
@@ -937,16 +1319,38 @@ export const completionSpec: Fig.Spec = {
...
     },
     {
       name: "mb",
       description: "Creates an S3 bucket.",
+      options: [
+        {
+          name: "--region",
+          description: "AWS region where the bucket is created",
+          args: {
+            name: "string",
```
Comment (cstrnt) [inline dev/aws/s3.ts:1345]:
```suggestion
            name: "region",
```

**Review by cstrnt — CHANGES_REQUESTED** [review]
Again, just a code review in terms of JS/TS Code, I have no clue about AWS 😆

File: dev/aws/s3.ts [hunk 15]
```diff
@@ -1,3 +1,346 @@
...
+    trigger: (newToken, oldToken) => {
+      return triggerPrefix(newToken, oldToken, _prefixS3);
+    },
+
+    filterTerm: (token) => {
+      if (!token.startsWith(_prefixS3)) return token;
+      return token.slice(token.lastIndexOf("/") + 1);
+    },
+    cache: {
+      ttl: 30000,
```
Comment (cstrnt) [inline dev/aws/s3.ts:294]: Same as the other PR with the top-level ttl :)

**Review by QuiiBz — APPROVED** [review]
Let's merge this big PR! Could you copy/paste the generators which are duplicated and reviewed in each AWS PR (#315, #297, #287, #244)?

File: dev/aws/s3.ts [hunk 16]
```diff
@@ -1,3 +1,368 @@
+// TODO: suggest available s3 endpoints
```
Comment (QuiiBz) [inline dev/aws/s3.ts:1]: Is this todo completed?

Comment (nagypeterjob) [reply dev/aws/s3.ts:1]: no, probably would be overkill. I don't think its a blocker

Comment (QuiiBz) [reply dev/aws/s3.ts:1]: Alright, thanks!

**Pull request merged**